cocos/animation/core/pose-allocator.ts

```diff
@@ -0,0 +1,148 @@
+import { assertIsTrue } from '../../core/data/utils/asserts';
+import { Pose } from './pose';
+import { TransformArray } from './transform-array';
+
+const MAX_POSE_PER_PAGE = 8;
+
+export class PoseAllocator {
+    constructor (transformCount: number, metaValueCount: number) {
+        this._transformCount = transformCount;
+        this._metaValueCount = metaValueCount;
+    }
+
+    public get allocatedCount () {
+        return this._allocatedCount;
+    }
+
+    public allocatePose (): Pose {
+        ++this._allocatedCount;
+        const { _pages: pages } = this;
+        const nPages = pages.length;
+        for (let iPage = 0; iPage < nPages; ++iPage) {
+            const page = pages[iPage];
+            const pose = page.tryAllocate();
+            if (pose) {
+                pose.pageIndex = iPage;
+                return pose;
+            }
+        }
+        return this._allocatePoseInNewPage();
+    }
+
+    public destroyPose (pose: Pose) {
+        assertIsTrue(pose instanceof PagedPose);
+        const { _pages: pages } = this;
+        const nPages = pages.length;
+        const pageIndex = pose.pageIndex;
+        assertIsTrue(pageIndex >= 0 && pageIndex < nPages);
+        const page = pages[pageIndex];
+        page.deallocate(pose);
+        --this._allocatedCount;
+    }
+
+    private _transformCount = 0;
+
+    private _metaValueCount = 0;
+
+    private _pages: PosePage[] = [];
+
+    private _allocatedCount = 0;
+
+    private _allocatePoseInNewPage (): PagedPose {
+        const page = new PosePage(this._transformCount, this._metaValueCount, 4);
+        const pageIndex = this._pages.length;
+        this._pages.push(page);
+        const pose = page.tryAllocate();
+        assertIsTrue(pose); // Shall not fail
+        pose.pageIndex = pageIndex;
+        return pose;
+    }
+}
+
+class PagedPose extends Pose {
```

Comment:
This type is a bit wired. Business class should not be coupled with allocator,  there may be various types of allocator in the future. I think it's better to keep a `memoryId` or `allocatorId` field in the `Pose` class to be set by the allocator, and all the rest of the logic is handled outside the pose.

Reply:
See https://github.com/cocos/cocos-engine/pull/13503/commits/c5e7a19125bc652c5b3f52d0552a9d8b64f48ac9 for my refactor